lib/knapsack_pro/runners/queue/rspec_runner.rb

```diff
@@ -3,224 +3,287 @@
+
+          @function_core.ensure_no_deprecated_run_all_when_everything_filtered_option!(@rspec_runner.knapsack__deprecated_run_all_when_everything_filtered_enabled?)
         end
 
-        # Clear rspec examples without the shared examples:
-        # https://github.com/rspec/rspec-core/pull/2379
-        #
-        # Keep formatters and report to accumulate info about failed/pending tests
-        def self.rspec_clear_examples
-          if ::RSpec::ExampleGroups.respond_to?(:remove_all_constants)
-            ::RSpec::ExampleGroups.remove_all_constants
-          else
-            ::RSpec::ExampleGroups.constants.each do |constant|
-              ::RSpec::ExampleGroups.__send__(:remove_const, constant)
-            end
-          end
-          ::RSpec.world.example_groups.clear
-          ::RSpec.configuration.start_time = ::RSpec::Core::Time.now
+        def post_run_tasks(exit_code)
+          @adapter_class.verify_bind_method_called
 
-          if KnapsackPro::Config::Env.rspec_split_by_test_examples?
-            # Reset example group counts to ensure scoped example ids in metadata
-            # have correct index (not increased by each subsequent run).
-            # Solves this problem: https://github.com/rspec/rspec-core/issues/2721
-            ::RSpec.world.instance_variable_set(:@example_group_counts_by_spec_file, Hash.new(0))
-          end
+          printable_args = @function_core.args_with_seed_option_added_when_viable(@rspec_runner.knapsack__seed_used?, @rspec_runner.knapsack__seed, @cli_args)
+          @function_core.log_rspec_command(printable_args, @node_assigned_test_file_paths, :end_of_queue)
 
-          # skip reset filters for old RSpec versions
-          if ::RSpec.configuration.respond_to?(:reset_filters)
-            ::RSpec.configuration.reset_filters
-          end
+          time_tracker = KnapsackPro::Formatters::TimeTrackerFetcher.call
+          KnapsackPro::Report.save_node_queue_to_api(time_tracker&.queue(@node_assigned_test_file_paths))
+
+          Kernel.exit(exit_code)
         end
 
-        def self.args_with_seed_option_added_when_viable(args, rspec_runner)
-          order_option = adapter_class.order_option(args)
+        def pull_tests_from_queue(can_initialize_queue: false)
+          test_file_paths = test_file_paths(
+            can_initialize_queue: can_initialize_queue,
+            executed_test_files: @node_assigned_test_file_paths
+          )
+          @node_assigned_test_file_paths += test_file_paths
+          test_file_paths
+        end
 
-          if order_option
-            # Don't add the seed option for order other than random, e.g. `defined`
-            return args unless order_option.include?('rand')
-            # Don't add the seed option if the seed is already set in args, e.g. `rand:12345`
-            return args if order_option.to_s.split(':')[1]
-          end
+        def with_queue_hooks(test_file_paths)
```

Comment:
It seems a bit out of place to see the ENV setting and wants to quit stuff in here given the name of the method.

Reply:
We could rename it to something like: `with_wrapped_batch`.

Reply:
"wrapped" doesn't add any concrete meaning.
 
I'd move everything excect the following back into the caller:

```ruby
          KnapsackPro::Hooks::Queue.call_before_subset_queue
          yield test_file_paths
          KnapsackPro::Hooks::Queue.call_after_subset_queue
```

and then extract one method for the env and one method for the quitting if needed.

I don't even know if we want to have a method just for the three lines tbh.

Reply:
The `with_queue_hooks` method exists to ensure the `with_batched_tests_from_queue` method (the caller) is only about calling the Queue API to get more tests. 

Is that adding any new information for you that helps design the structure?
I need to think more about what to do with `with_queue_hooks` because I'm a bit lost.

Reply:
Not sure I understand the question.

My point is that `with_queue_hooks` is doing more than just calling the hooks. I'd get rid of `with_queue_hooks` and inline the code in `with_batched_tests_from_queue`:

```ruby
def with_batched_tests_from_queue
  loop do
      # 1
      subset_queue_id = KnapsackPro::Config::EnvGenerator.set_subset_queue_id
      ENV['KNAPSACK_PRO_SUBSET_QUEUE_ID'] = subset_queue_id

      # 2
      KnapsackPro::Hooks::Queue.call_before_subset_queue
      yield test_file_paths
      KnapsackPro::Hooks::Queue.call_after_subset_queue

      # 3
      if @rspec_runner.knapsack__wants_to_quit?
        KnapsackPro.logger.warn('RSpec wants to quit.')
        set_terminate_process
      end
      if @rspec_runner.knapsack__rspec_is_quitting?
        KnapsackPro.logger.warn('RSpec is quitting.')
        set_terminate_process
      end
      printable_args = @function_core.args_with_seed_option_added_when_viable(@rspec_runner.knapsack__seed_used?, @rspec_runner.knapsack__seed, @cli_args)
      @function_core.log_rspec_command(printable_args, test_file_paths, :subset_queue)
  end
end
```

You may then want to extract functions or not. Prolly \#3 would benefit from it.

Reply:
after refactor: https://github.com/KnapsackPro/knapsack_pro-ruby/pull/237/commits/c0e8745581f15050329d207940c70b96e9bb02d3

I decided the `with_batch` method is going to do more than just API calls as you proposed in the above example.

I extracted log rspec methods.